---
title: "<center> The COVID challenge <center>"
author: "<center> Sébastien Wieckowski <center><br>"
date: "<center> _`r Sys.Date()`_ <center>"
output:
  pdf_document:
    fig_caption: yes
    toc: yes
    latex_engine: xelatex
---

```{r setup, include=FALSE}
library(rmarkdown)
library(tinytex)
library(knitr)
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE)
```

## Introduction

The proposed challenge is about the exploitation of raw data for an extensive RNA-seq exploration of the host response to SARS-CoV-2 and related viruses in different cell lines, tissues and time points, and which include human primary samples as well as _in vivo_ studies in ferrets - an incredible dataset from Benjamin TenOever’s lab at Mt Sinai! You should check out their recent BioRxiv preprint: [SARS-CoV-2 launches a unique transcriptional signature from in vitro, ex vivo, and in vivo systems](https://doi.org/10.1101/2020.03.24.004655).  

### Study design

We first looked at the original study design file, in particular by selecting and sorting the sample, cell_line, treatment (which contains the MOI) and timepoint_postTreatment columns. The 'messy' study design reports a variety of different samples and conditions, including:

* human primary NHBE samples treated mock vs. SARS-CoV-2 (from now SARS), IAV (i.e. influenza A virus), IAVdNS1 or IFN-β for different timepoints (12 - 24 h)
* human A549 cell line treated mock vs. SARS (MOIs 0.2 or 2), RSV (MOIs 2 or 15) or HIPV3 (MOI 3) for 24 h
* human A549 cells treated mock vs. IAV (MOI 5) for 9 h
* human A549-ACE2 treated mock vs. SARS (MOIs 0.2, 2) or pretreated with Ruxolitinib and then SARS (MOI 2) for 24 h
* human Calu3 cell line treated mock vs. SARS (MOI 2) for 24 h
* _in vivo_ nasal wash and trachea (only one time point) samples from ferrets treated mock vs. SARS (5E4 PFU) for 1, 3, 7 or 14 days, and IAV (1E5 PFU) at day 7 only
* some other lung biopsies from patients

While no gene ontology is available for furret, the time course in the _in vivo_ experiment and the comparison with _ex vivo_ data using primary normal human bronchial epithelia (NHBE) cells seem to be an interesting starting point for multivariate analysis of the gene signatures regulated during the infection with SARS-CoV-2. Moreover we can compare the responses to the ones induced after infection with influenza A virus.  
The plan for the challenge is therefore:

1. classical RNAseq data analysis including filtering, normalization and PCA in both the human primary cells infected with SARS or IAV, and ferret's nasal wash samples
2. DEG in human primary cells up to heatmap and functional enrichment analysis
3. module identification in multiple datasets defined as mock, SARS and IAV during the time course of the infection using `clust`

***

## R packages used

All graphics and data wrangling were handled using the [tidyverse suite of packages](https://www.tidyverse.org/). All packages used are available from the Comprehensive R Archive Network (CRAN), Bioconductor.org, or Github.

***

## Read the data

The goal of this short script to it make sure we are able to read the count data and the study design for the COVID hackdash into the R environment using tidyverse. Both the human and ferret expression data have been obtained as **raw count tables** from the public GEO entry (see challenge description for more details), therefore no annotation is required.

```{r data import}
library(tidyverse)

# Reading in study design **as tibble** incl all info for both human and ferret samples
targets <- read_tsv("covid_metadata.txt")

# human covid data read/converted to a matrix with gene symbols as rownames
# initially the dataframe contains X1 col with the name of genes
human_covid_data <- read_tsv("GSE147507_RawReadCounts_Human.tsv")
human_covid_data <- as.matrix(column_to_rownames(human_covid_data, "X1"))

# repeat for the ferret covid data
ferret_covid_data <- read_tsv("GSE147507_RawReadCounts_Ferret.tsv")
ferret_covid_data <- as.matrix(column_to_rownames(ferret_covid_data, "X1"))
# then the datatype is homogeneous in the matrices
```

When we look at the variables of the targets dataframe: `r colnames(targets)`, it seems straightforward to subset data from human and ferret experiments since unique elements from the `host_common_name` variable are `r unique(targets$host_common_name)`.

### Data filtering

We then prepare and subset data we are interested in for this analysis. In the human experiment we keep the data from primary NHBE samples infected with mock, SARS or IAV (not the other conditions) thanks to the elements in the `tissue_cell_type` variable: `r unique(targets$tissue_cell_type)`. Of note we pool together the data of mock-treated cells after 12 **and** 24 hours.

```{r data subsetting human}
targets_human <- targets %>%
  dplyr::select(sample, host_common_name, tissue_cell_type, treatment) %>%
  # timepoint not used here
  dplyr::filter(host_common_name == "human" &
                  tissue_cell_type == "primary human bronchial epithelial cells" &
                  treatment %in% c("Mock treatment", "SARS-CoV-2 infected (MOI 2)",
                                 "IAV infected (MOI 3)"))

# then we don't need the host_common_name and tissue_cell_type anymore
targets_ferret <- targets_human %>%
  dplyr::select(sample, treatment)

treatment_human <- factor(targets_human$treatment,
                           levels = c("Mock treatment", "SARS-CoV-2 infected (MOI 2)",
                                      "IAV infected (MOI 3)"),
                           labels = c("ctrl", "SARS", "IAV"))

# and finally we keep only the columns of interest in the count matrix
human_data <- human_covid_data[, targets_human$sample]

```

Before subsetting the human data, the dimension of the original human_covid_data was `r dim(human_covid_data)`and after cleaning, the dimension of the human_data is `r dim(human_data)`.

In the _in vivo_ study, all the data with nasal wash sample type are kept thanks to the elements in the `tissue_cell_type` variable: `r unique(targets$tissue_cell_type)`.

```{r data subsetting ferret}
# we first select the columns of interest and further filter of ferret data
targets_ferret <- targets %>%
  dplyr::select(sample, host_common_name, tissue_cell_type, treatment, timepoint_postTreament) %>% # beware the typo in 'Treament'
  dplyr::filter(host_common_name == "ferret" & tissue_cell_type == "Nasal Wash")

# then we don't need the host_common_name and tissue_cell_type anymore
targets_ferret <- dplyr::select(targets_ferret, sample, treatment, timepoint_postTreament)
colnames(targets_ferret) <- c("sample", "treatment", "time")

treatment_ferret <- factor(targets_ferret$treatment,
                           levels = c("Mock treatment", "SARS-CoV-2 infected (5E4 PFU)",
                                      "IAV infected (1E5 PFU)"),
                           labels = c("ctrl", "SARS", "IAV"))
time_ferret <- factor(targets_ferret$time,
                      levels = c("day1", "day3", "day7", "day14"),
                      labels = c(1, 3, 7, 14))

# and finally we keep only the columns of interest in the count matrix
ferret_data <- ferret_covid_data[, targets_ferret$sample]
```

Before subsetting the ferret data, the dimension of the original ferret_covid_data was `r dim(ferret_covid_data)`and after cleaning, the dimension of the ferret_data is `r dim(ferret_data)`.

***

## Preprocessing human dataset

### Impact of filtering and normalization

```{r function profile, echo=TRUE}
library(edgeR)
library(matrixStats)

# the function below is used for ploting violins
profile_function <- function(data, samples, title, subtitle) {
    data_count <- edgeR::cpm(data, log = TRUE)
    data_count_df <- tibble::as_tibble(data_count, rownames = "geneID")
    colnames(data_count_df) <- c("geneID", samples)
    data_count_df_pivot <- tidyr::pivot_longer(data_count_df,
                                               cols = -1, # all except col 1
                                               names_to = "samples",
                                               values_to = "expression")

    ggplot(data_count_df_pivot) +
      aes(x = samples, y = expression, fill = samples) +
      geom_violin(trim = FALSE, show.legend = FALSE) +
      stat_summary(fun = "median",
                   geom = "point",
                   shape = 95,
                   size = 10,
                   color = "black",
                   show.legend = FALSE) +
      labs(y = "log2 expression", x = "sample",
           title = title,
           subtitle = subtitle) +
      theme_bw() +
      theme(axis.text.x=element_blank(), # hide x tick labels
            plot.title = element_text(size=12))
}
```

```{r data wrangling human dataset}
library(cowplot)
human_sample_labels <- targets_human$sample
human_DGEList <- DGEList(human_data)
# we take advantage of the profile_function that return violin_plot of cpm
human_p1 <- profile_function(data = human_DGEList,
                             samples = human_sample_labels,
                             title = "human dataset",
                             subtitle = "unfiltered, non-normalized")

human_cpm <- cpm(human_DGEList)
human_keepers <- rowSums(human_cpm > 1) >= 3 # there are triplic/quadrupl-ates
human_DGEList_filtered <- human_DGEList[human_keepers, ]
human_p2 <- profile_function(data = human_DGEList_filtered,
                             samples = human_sample_labels,
                             title = "human dataset",
                             subtitle = "filtered, non-normalized")

human_DGEList_filtered_norm <- calcNormFactors(human_DGEList_filtered,
                                               method = "TMM")
human_p3 <- profile_function(data = human_DGEList_filtered_norm,
                             samples = human_sample_labels,
                             title = "human dataset",
                             subtitle = "filtered, normalized")

plot_grid(human_p1, human_p2, human_p3,
          labels = c('A', 'B', 'C'),
          label_size = 12)
```

Filtering was carried out to remove lowly expressed genes. Genes with less than 1 count per million (CPM) in at least 5 or more samples filtered out.  This reduced the number of genes from `r nrow(human_DGEList)` to `r nrow(human_DGEList_filtered)`.

***

## PCA plot

Time to see if infection with SARS-CoV-2 induces a particular gene expression signature, and if this one is different after infection with IAV. The treatment groups have been already factorized, see: `r treatment_human`. _Note: As we don't really need any table for the time being, no datatable is produced in this document._

```{r PCA plot human dataset}
human_pca_res <- prcomp(t(data.matrix(human_DGEList_filtered_norm)), scale.=F, retx=T)
human_pc_var <- human_pca_res$sdev^2
human_pc_per <- round(human_pc_var/sum(human_pc_var) * 100, 1)
human_pca_res_df <- as_tibble(human_pca_res$x)

ggplot(human_pca_res_df) +
  aes(x = PC1, y = PC2, label = human_sample_labels, color = treatment_human) +
  geom_point(size = 4) +
  stat_ellipse() +
  xlab(paste0("PC1 (", human_pc_per[1], "%", ")")) +
  ylab(paste0("PC2 (", human_pc_per[2], "%", ")")) +
  labs(title = "PCA plot") +
  coord_fixed() +
  theme_bw()
```
This is not super convincing although PC1 and PC2 account for `r paste0(human_pc_per[1] + human_pc_per[2], "%")` of the total variance.

***

## Differentially expressed genes

To identify differentially expressed genes, precision weights were first applied to each gene based on its mean-variance relationship using [VOOM](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29), then data was normalized using the [TMM method](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25) in [EdgeR](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/).  Linear modeling and bayesian stats were employed via [Limma](https://academic.oup.com/nar/article/43/7/e47/2414268) to find genes that were up- or down-regulated in leishmania patients by 4-fold or more, with a false-discovery rate (FDR) of 0.01.

### Volcano plot

Let's see if there is any interesting gene differentially expressed detected between SARS-CoV-2 and mock conditions.

```{r differential genes human dataset (volcano plot)}
library(limma)
library(edgeR)
library(gt)

design <- model.matrix(~0 + treatment_human)
colnames(design) <- levels(treatment_human)

v_human_DGEList_filtered_norm <- voom(human_DGEList_filtered_norm, design, plot = F)
human_fit <- lmFit(v_human_DGEList_filtered_norm, design)
human_contrast_matrix <- makeContrasts(infection = SARS - ctrl,
                                       levels = design)

human_fits <- contrasts.fit(human_fit, human_contrast_matrix)
human_ebFit <- eBayes(human_fits)
human_top_hits <- topTable(human_ebFit,
                           adjust = "BH",
                           coef = 1,
                           number = 20000,
                           sort.by = "logFC")

human_top_hits_df <- human_top_hits %>%
  as_tibble(rownames = "geneID")

ggplot(human_top_hits_df) +
  aes(y = -log10(adj.P.Val), x = logFC, text = paste("Symbol:", geneID)) +
  geom_point(size = 2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "Infection with SARS-CoV-2 vs. mock (24 h)") +
  theme_bw()
```
From the volcano plot, it seems that a dozen of genes are upregulated after infection by SARS-CoV-2.

***

### Table of DEGs

```{r differential genes human dataset (DEG table)}
human_results <- decideTests(human_ebFit,
                             method = "global",
                             adjust.method = "BH",
                             p.value = 0.05,
                             lfc = 1)
colnames(v_human_DGEList_filtered_norm$E) <- human_sample_labels
human_diffGenes <- v_human_DGEList_filtered_norm$E[human_results[, 1] !=0, ]
human_diffGenes_df <- as_tibble(human_diffGenes, rownames = "geneID")

human_diffGenes_df %>%
  sample_n(10) %>%
  gt() %>%
  fmt_number(columns = 2:11, decimals = 2)
```
Great to see IL-6 in the list, but nothing obvious otherwise...

***

## Heatmaps and modules

Pearson correlation was used to cluster **`r nrow(human_diffGenes)`** differentially expressed genes, which were then represented as heatmap with the data scaled by Zscore for each row.

```{r modules human dataset (heatmap)}
library(gplots)
library(RColorBrewer)

myheatcolors <- rev(brewer.pal(name="RdBu", n=11))

clustRows <- hclust(as.dist(1 - cor(t(human_diffGenes),
                                    method = "pearson")),
                    method = "complete")
clustColumns <- hclust(as.dist(1 - cor(human_diffGenes,
                                       method="spearman")),
                       method="complete")

module.assign <- cutree(clustRows, k = 2)
module.color <- rainbow(length(unique(module.assign)), start=0.1, end=0.9)
module.color <- module.color[as.vector(module.assign)]
heatmap.2(human_diffGenes,
          Rowv = as.dendrogram(clustRows),
          Colv = as.dendrogram(clustColumns),
          RowSideColors = module.color,
          col = myheatcolors, scale = 'row', labRow = NA,
          density.info = "none", trace = "none",
          cexRow = 1, cexCol = 1, margins = c(8,8))
```
It looks like most of the genes are upregulated upon infection with SARS-CoV-2. Interestinlgy infection with IAV also lead to a similar profile but to a lesser extent. We then focus the rest of the analysis with upregulated genes. Of note, some replicate in the IAV treatment group have kind of opposite profile, which fits well with the PCA analysis.

```{r modules human dataset (upregulated genes)}
modulePick <- 2
myModule_up <- human_diffGenes[names(module.assign[module.assign %in% modulePick]), ]
hrsub_up <- hclust(as.dist(1 - cor(t(myModule_up),
                                   method = "pearson")), 
                   method = "complete")

heatmap.2(myModule_up,
          Rowv = as.dendrogram(hrsub_up),
          Colv = NA,
          labRow = NA,
          col = myheatcolors, scale = "row",
          density.info = "none", trace = "none",
          RowSideColors = module.color[module.assign %in% modulePick],
          margins=c(10,8))
```

## GO enrichment

GO enrichment for the `r nrow(myModule_up)` genes induced by infection with SARS-Cov-2

```{r functional enrichment human dataset (gostplot for upregulated genes)}
library(GSEABase) # functions and methods for Gene Set Enrichment Analysis
library(Biobase) # base functions for bioconductor; required by GSEABase
library(GSVA) # GSVA, a non-parametric and unsupervised method for estimating variation of gene set enrichment across samples
library(gprofiler2) # tools for accessing the GO enrichment results using g:Profiler web resources
library(clusterProfiler) # provides a suite of tools for functional enrichment analysis
gost.res_up <- gost(rownames(myModule_up),
                    organism = "hsapiens",
                    correction_method = "fdr")
gostplot(gost.res_up,
         interactive = F,
         capped = F) # set interactive=FALSE to get plot for publications
```
Most of the upregulated genes lies within the **cellular component** ontology group. It might also be interesting to see more closely what are the genes in the WP group.

## GSEA

```{r functional enrichment human dataset (GSEA table)}
library(msigdbr) # access to msigdb collections directly within R
library(enrichplot) # great for making the standard GSEA enrichment plots

hs_gsea_c2 <- msigdbr(species = "Homo sapiens",
                      category = "C2") %>%
  dplyr::select(gs_name, gene_symbol) # just get the columns corresponding to signature name and gene symbols of genes in each signature

# Now that you have your msigdb collections ready, prepare your data
# Pull out just the columns corresponding to gene symbols and LogFC for at least one pairwise comparison for the enrichment analysis
log2.cpm.filtered.norm.df <- cpm(human_DGEList_filtered_norm, log = TRUE) %>%
  as_tibble(rownames = "geneID")
# needs to rename columns
colnames(log2.cpm.filtered.norm.df) <- c("geneID",
                                         c("mock_1", "mock_2", "mock_3",
                                           "sars_1", "sars_2", "sars_3",
                                           "mock_4", "mock_5", "mock_6", "mock_7",
                                           "iav_1", "iav_2", "iav_3", "iav_4"))
mydata.df <- log2.cpm.filtered.norm.df  %>%
  mutate(mock.AVG = (mock_1 + mock_2 + mock_3 + mock_4 + mock_5 + mock_6 + mock_7) / 7,
         infect.AVG = (sars_1 + sars_2 + sars_3) / 3,
         #now make columns comparing each of the averages above that you're interested in
         LogFC = (infect.AVG - mock.AVG)) %>%
  mutate_if(is.numeric, round, 2)
# Pull out just the columns corresponding to gene symbols and LogFC for at least one pairwise comparison for the enrichment analysis
mydata.df.sub <- dplyr::select(mydata.df, geneID, LogFC)
# construct a named vector
mydata.gsea <- mydata.df.sub$LogFC
names(mydata.gsea) <- as.character(mydata.df.sub$geneID)
mydata.gsea <- sort(mydata.gsea, decreasing = TRUE)


mydata.df.sub <- dplyr::select(mydata.df, geneID, LogFC)
mydata.gsea <- mydata.df.sub$LogFC
names(mydata.gsea) <- as.character(mydata.df.sub$geneID)
mydata.gsea <- sort(mydata.gsea, decreasing = TRUE)

# run GSEA using the 'GSEA' function from clusterProfiler
myGSEA.res <- GSEA(mydata.gsea,
                   TERM2GENE = hs_gsea_c2,
                   verbose=FALSE)
myGSEA.df <- as_tibble(myGSEA.res@result)

# view results as a table
myGSEA.df %>%
  dplyr::arrange(NES) %>%
  sample_n(10) %>%
  gt()
```

```{r functional enrichment human dataset IL-6 pathway (enrich plot)}
# create enrichment plots using the enrichplot package
gseaplot2(myGSEA.res,
          geneSetID = 21, # IL-6 signaling pathway
          pvalue_table = FALSE, #can set this to FALSE for a cleaner plot
          title = "IL-6 signaling pathway")
```

```{r functional enrichment human dataset (bubble plot)}
# add a variable to this result that matches enrichment direction with phenotype
myGSEA.df <- myGSEA.df %>%
  mutate(phenotype = case_when(
    NES > 0 ~ "infected",
    NES < 0 ~ "control"))

# create 'bubble plot' to summarize y signatures across x phenotypes
ggplot(myGSEA.df[1:20,],
       aes(x=phenotype, y=ID)) +
  geom_point(aes(size=setSize,
                 color = NES,
                 alpha = -log10(p.adjust))) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw()
```

***

## Preprocessing ferret dataset

### Impact of filtering and normalization

```{r data wrangling ferret dataset}
ferret_sample_labels <- targets_ferret$sample
ferret_DGEList <- DGEList(ferret_data)
# we take advantage of the profile_function that return violin_plot of cpm
ferret_p1 <- profile_function(data = ferret_DGEList,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "unfiltered, non-normalized")

ferret_cpm <- cpm(ferret_DGEList)
ferret_keepers <- rowSums(ferret_cpm > 10) >= 2 # all duplicates
ferret_DGEList_filtered <- ferret_DGEList[ferret_keepers, ]
ferret_p2 <- profile_function(data = ferret_DGEList_filtered,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "filtered, non-normalized")

ferret_DGEList_filtered_norm <- calcNormFactors(ferret_DGEList_filtered,
                                               method = "TMM")
ferret_p3 <- profile_function(data = ferret_DGEList_filtered_norm,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "filtered, normalized")

plot_grid(ferret_p1, ferret_p2, ferret_p3,
          labels = c('A', 'B', 'C'),
          label_size = 12)
```

Filtering was carried out to remove lowly expressed genes. The threshold count is somehow higher than for human data, and does not filter all the lowly expressed genes, but looking at the non-normalized data it seems important not to increase the threshold in order to detetct relatively lowly expressed genes in some groups. Genes with less than 1 count per million (CPM) in at least 5 or more samples filtered out.  This reduced the number of genes from `r nrow(ferret_DGEList)` to `r nrow(ferret_DGEList_filtered)`.

***

## PCA plot ferret data

Time to see if infection with SARS-CoV-2 induces a particular gene expression signature, and if this one is different after infection with IAV. The treatment groups have been already factorized, see: `r treatment_ferret`. The idea here is to look at the PCA and create module for clust.

```{r PCA plot ferret dataset}
ferret_pca_res <- prcomp(t(data.matrix(ferret_DGEList_filtered_norm)), scale.=F, retx=T)
ferret_pc_var <- ferret_pca_res$sdev^2
ferret_pc_per <- round(ferret_pc_var/sum(ferret_pc_var) * 100, 1)
ferret_pca_res_df <- as_tibble(ferret_pca_res$x)

ggplot(ferret_pca_res_df) +
  aes(x = PC1, y = PC2, label = ferret_sample_labels,
      color = time_ferret,
      shape = treatment_ferret) +
  geom_point(size = 4) +
  #stat_ellipse() +
  xlab(paste0("PC1 (", ferret_pc_per[1], "%", ")")) +
  ylab(paste0("PC2 (", ferret_pc_per[2], "%", ")")) +
  labs(title = "PCA plot",
       subtitle = "ferret dataset") +
  coord_fixed() +
  theme_bw()
```
This is not easy to interpret although PC1 and PC2 account for `r paste0(ferret_pc_per[1] + ferret_pc_per[2], "%")` of the total variance. It seems that no difference is detected at earlier time points but only between control and SARS groups at day 14 post-infection, though duplicates are spread. Let's have a look at other PCs.

```{r PCA ferret multiple charts}
pca.res.df <- ferret_pca_res$x[, 1:4] %>%
  as_tibble() %>%
  add_column(sample = ferret_sample_labels,
             treatment = treatment_ferret)
  
pca.pivot <- pivot_longer(pca.res.df,
                          cols = PC1:PC4,
                          names_to = "PC",
                          values_to = "loadings")

ggplot(pca.pivot) +
  aes(x = sample, y = loadings, fill = treatment) +
  geom_bar(stat = "identity") +
  facet_wrap(~PC) +
  labs(title = "PCA 'small multiples' plot",
       subtitle = "ferret dataset") +
  theme_bw() +
  coord_flip()
```
That is really weird, one sample from the control group at day 14 seems to belong to the SARS group.

***

## Differentially expressed genes ferret

Let's see if there is any interesting gene differentially expressed detected between SARS-CoV-2 and mock conditions.

```{r diffGenes ferret dataset (volcano plot)}
library(gt)

design <- model.matrix(~0 + treatment_ferret)
colnames(design) <- levels(treatment_ferret)

v_ferret_DGEList_filtered_norm <- voom(ferret_DGEList_filtered_norm, design, plot = T)
ferret_fit <- lmFit(v_ferret_DGEList_filtered_norm, design)
ferret_contrast_matrix <- makeContrasts(infection = SARS - ctrl,
                                        levels = design)

ferret_fits <- contrasts.fit(ferret_fit, ferret_contrast_matrix)
ferret_ebFit <- eBayes(ferret_fits)

ferret_results <- decideTests(ferret_ebFit,
                              method = "global",
                              adjust.method = "BH",
                              p.value = 0.01,
                              lfc = 1)
colnames(v_ferret_DGEList_filtered_norm$E) <- ferret_sample_labels
ferret_diffGenes <- v_ferret_DGEList_filtered_norm$E[ferret_results[, 1] !=0, ]
ferret_diffGenes_df <- as_tibble(ferret_diffGenes, rownames = "geneID")

#ferret_diffGenes_df %>%
#  sample_n(10) %>%
#  gt() %>%
#  fmt_number(columns = 2:11, decimals = 2)
```

The list is empty, let's check on a volcano plot.

```{r differential genes ferret dataset (volcano plot)}
ferret_top_hits <- topTable(ferret_ebFit,
                            adjust = "BH",
                            coef = 1,
                            number = 20000,
                            sort.by = "logFC")

ferret_top_hits_df <- ferret_top_hits %>%
  as_tibble(rownames = "geneID")

ggplot(ferret_top_hits_df) +
  aes(y = -log10(adj.P.Val), x = logFC, text = paste("Symbol:", geneID)) +
  geom_point(size = 2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "Infection with SARS-CoV-2 vs. mock (24 h)") +
  theme_bw()
```
The analysis looks compromised for this timepoint as not significant logFC are detected, and we won't produce a clust. Let's filter this latest time point and redo all analyses for the ferret dataset.

```{r data subsetting ferret 2}
# we first select the columns of interest and further filter of ferret data
targets_ferret <- targets %>%
  dplyr::select(sample, host_common_name, tissue_cell_type, treatment, timepoint_postTreament) %>% # beware the typo in 'Treament'
  dplyr::filter(host_common_name == "ferret" & tissue_cell_type == "Nasal Wash" & timepoint_postTreament != "day14")

# then we don't need the host_common_name and tissue_cell_type anymore
targets_ferret <- dplyr::select(targets_ferret, sample, treatment, timepoint_postTreament)
colnames(targets_ferret) <- c("sample", "treatment", "time")

treatment_ferret <- factor(targets_ferret$treatment,
                           levels = c("Mock treatment", "SARS-CoV-2 infected (5E4 PFU)",
                                      "IAV infected (1E5 PFU)"),
                           labels = c("ctrl", "SARS", "IAV"))
time_ferret <- factor(targets_ferret$time,
                      levels = c("day1", "day3", "day7"),
                      labels = c(1, 3, 7))

# and finally we keep only the columns of interest in the count matrix
ferret_data <- ferret_covid_data[, targets_ferret$sample]
```

```{r data wrangling ferret dataset 2}
ferret_sample_labels <- targets_ferret$sample
ferret_DGEList <- DGEList(ferret_data)
# we take advantage of the profile_function that return violin_plot of cpm
ferret_p1 <- profile_function(data = ferret_DGEList,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "unfiltered, non-normalized")

ferret_cpm <- cpm(ferret_DGEList)
ferret_keepers <- rowSums(ferret_cpm > 10) >= 2 # all duplicates
ferret_DGEList_filtered <- ferret_DGEList[ferret_keepers, ]
ferret_p2 <- profile_function(data = ferret_DGEList_filtered,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "filtered, non-normalized")

ferret_DGEList_filtered_norm <- calcNormFactors(ferret_DGEList_filtered,
                                               method = "TMM")
ferret_p3 <- profile_function(data = ferret_DGEList_filtered_norm,
                              samples = ferret_sample_labels,
                              title = "ferret dataset",
                              subtitle = "filtered, normalized")

plot_grid(ferret_p1, ferret_p2, ferret_p3,
          labels = c('A', 'B', 'C'),
          label_size = 12)
```

```{r PCA plot ferret dataset 2}
ferret_pca_res <- prcomp(t(data.matrix(ferret_DGEList_filtered_norm)), scale.=F, retx=T)
ferret_pc_var <- ferret_pca_res$sdev^2
ferret_pc_per <- round(ferret_pc_var/sum(ferret_pc_var) * 100, 1)
ferret_pca_res_df <- as_tibble(ferret_pca_res$x)

ggplot(ferret_pca_res_df) +
  aes(x = PC1, y = PC2, label = ferret_sample_labels,
      color = time_ferret,
      shape = treatment_ferret) +
  geom_point(size = 4) +
  #stat_ellipse() +
  xlab(paste0("PC1 (", ferret_pc_per[1], "%", ")")) +
  ylab(paste0("PC2 (", ferret_pc_per[2], "%", ")")) +
  labs(title = "PCA plot",
       subtitle = "ferret dataset") +
  coord_fixed() +
  theme_bw()
```

Now it looks better, and while PC1 and PC2 account for `r paste0(ferret_pc_per[1] + ferret_pc_per[2], "%")` of the total variance, we can see IAV infection induces a specific gene expression signature compared to SARS at day 7 post-infection. Interestingly, SARS differs to some extent to the control group at day 3.

```{r diffGenes ferret dataset (volcano plot) 2}
library(gt)

design <- model.matrix(~0 + treatment_ferret)
colnames(design) <- levels(treatment_ferret)

v_ferret_DGEList_filtered_norm <- voom(ferret_DGEList_filtered_norm, design, plot = T)
ferret_fit <- lmFit(v_ferret_DGEList_filtered_norm, design)
ferret_contrast_matrix <- makeContrasts(infection = SARS - ctrl,
                                        levels = design)

ferret_fits <- contrasts.fit(ferret_fit, ferret_contrast_matrix)
ferret_ebFit <- eBayes(ferret_fits)

ferret_results <- decideTests(ferret_ebFit,
                              method = "global",
                              adjust.method = "BH",
                              p.value = 0.01,
                              lfc = 1)
colnames(v_ferret_DGEList_filtered_norm$E) <- ferret_sample_labels
ferret_diffGenes <- v_ferret_DGEList_filtered_norm$E[ferret_results[, 1] !=0, ]
ferret_diffGenes_df <- as_tibble(ferret_diffGenes, rownames = "geneID")
```

```{r differential genes ferret dataset (volcano plot) 2}
ferret_top_hits <- topTable(ferret_ebFit,
                            adjust = "BH",
                            coef = 1,
                            number = 20000,
                            sort.by = "logFC")

ferret_top_hits_df <- ferret_top_hits %>%
  as_tibble(rownames = "geneID")

ggplot(ferret_top_hits_df) +
  aes(y = -log10(adj.P.Val), x = logFC, text = paste("Symbol:", geneID)) +
  geom_point(size = 2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "Infection with SARS-CoV-2 vs. mock (24 h)") +
  theme_bw()
```

Again no significantly different gene expression could be observe after infection with SARS.

***

## Conclusions

The data are not super convincing to me, no huge gene signature popped up upon infection with SARS-CoV-2. Moreover there is not really a good rationale to compare data from IAV group as the infection time are different. Last point was to detect modules from _in vivo_ experiment with ferret using the clust software. Because de differential expression gene analysis was not conclusive, we did not run clust.

Altogether, neither the _ex vivo_ experiment not the _in vivo_ data generated in ferret go in the direction of a dramatic gene signature in response to SARS-CoV-2 infection.

## Session info

The output from running 'sessionInfo' is shown below and details all packages and version necessary to reproduce the results in this report.

```{r session info}
sessionInfo()
```
